Dockerfile

```
@@ -1,23 +1,20 @@
+# defaulted ARGs need to be declared first
 ARG BUILDER_IMAGE=golang:1.24
 ARG BASE_IMAGE=gcr.io/distroless/static:nonroot
-
-# Build the manager binary
+# compilation stage for the manager binary
 FROM --platform=${BUILDPLATFORM} ${BUILDER_IMAGE} AS builder
-
-ARG CGO_ENABLED
-ARG TARGETARCH
-
 WORKDIR /workspace
-
-# Copy the go source
+# fetch dependencies first, for iterative development
+COPY go.mod go.sum ./
+RUN go mod download
+# copy the rest of the sources and build
 COPY . .
-
-# Build
+ARG CGO_ENABLED TARGETARCH
 RUN make build GO_BUILD_ENV='CGO_ENABLED=${CGO_ENABLED} GOOS=linux GOARCH=${TARGETARCH}'
```

Not sure why we didn't do:
```suggestion
RUN make build GO_BUILD_ENV="CGO_ENABLED=${CGO_ENABLED} GOOS=linux GOARCH=${TARGETARCH}"
```
To substitute the args earlier.

This currently does:
`RUN make build GO_BUILD_ENV='CGO_ENABLED=${CGO_ENABLED} GOOS=linux GOARCH=${TARGETARCH}'`

And then the makefile ultimately eventually expands these.

Switching to the suggestion above instead does:
`RUN make build GO_BUILD_ENV="CGO_ENABLED=0 GOOS=linux GOARCH=arm64"`
(docker subsitutes them early on)

Was there a reason for this? If not I think we should also fix that.
If nothing else it makes the build logs clearer.

---

I looked at the commits and found this one: https://github.com/kubernetes-sigs/kueue/commit/7cb85d24e83b42659ff73aa05be923ce3193055d#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557L18-R20

my guess is there was single quote before, then we parametrized, and it continued to work, so we just moved ahead, but cc @tenzen-y in case there is more to it.

---

I think we can remove `GOOS` and `GOARCH` from this build script which means `RUN make build GO_BUILD_ENV="CGO_ENABLED=${CGO_ENABLED}`. However, we might want to keep `CGO_ENABLED` since IIUC, some user want to handle CGO based on their policies.

Actually, we added CGO env based on their requests.

---

> I think we can remove GOOS and GOARCH 

Maybe, but I wouldn't require it form the PR which seems more oriented to best practices for writing the Dockerfiles. I think we can address that in a follow up issue / pr.

---

Yeah, that makes sense.
@BenTheElder Thank you!

/lgtm
/approve

---

We can't remove GOARCH because we are cross compiling, but we can drop GOOS, since it will always be linux in the container. https://github.com/kubernetes-sigs/kueue/pull/5205